pkg/scheduler/priorities.go

```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
 	}
 	return result, nil
 }
+
+// BalancedResourceAllocation is favors nodes with balanced resource usage rate.
+// BalancedResourceAllocation should **NOT** be used alone, and **MUST** be used together with LeastRequestedPriority.
+// It calculates the percentage of memory and CPU requested over capacity, and prioritizes based on how close the
+// two percentage metrics are to each other.
+// Details: score = 1 / sqrt[(cpu_pct - average)^2 + (mem_pct - average)^2], average = (cpu_pct + mem_pct) / 2
+func BalancedResourceAllocation(pod api.Pod, podLister PodLister, minionLister MinionLister) (HostPriorityList, error) {
+	nodes, err := minionLister.List()
+	if err != nil {
+		return HostPriorityList{}, err
+	}
+	podsToMachines, err := MapPodsToMachines(podLister)
+
+	list := HostPriorityList{}
+	for _, node := range nodes.Items {
+		list = append(list, calculateBalancedResourceAllocation(pod, node, podsToMachines[node.Name]))
+	}
+	return list, nil
+}
+
+func calculateBalancedResourceAllocation(pod api.Pod, node api.Node, pods []api.Pod) HostPriority {
+	totalMilliCPU := int64(0)
+	totalMemory := int64(0)
+	score := int(0)
+	for _, existingPod := range pods {
+		for _, container := range existingPod.Spec.Containers {
+			totalMilliCPU += container.Resources.Limits.Cpu().MilliValue()
+			totalMemory += container.Resources.Limits.Memory().Value()
+		}
+	}
+	// Add the resources requested by the current pod being scheduled.
+	// This also helps differentiate between differently sized, but empty, minions.
+	for _, container := range pod.Spec.Containers {
+		totalMilliCPU += container.Resources.Limits.Cpu().MilliValue()
+		totalMemory += container.Resources.Limits.Memory().Value()
+	}
+
+	capacityMilliCPU := node.Status.Capacity.Cpu().MilliValue()
+	capacityMemory := node.Status.Capacity.Memory().Value()
+
+	cpuPctOfCapRequested := calculatePctOfCapRequested(totalMilliCPU, capacityMilliCPU, node.Name)
+	memoryPctOfCapRequested := calculatePctOfCapRequested(totalMemory, capacityMemory, node.Name)
+	avergePctOfCapRequested := (cpuPctOfCapRequested + memoryPctOfCapRequested) / 2
+	cpuDiff := cpuPctOfCapRequested - avergePctOfCapRequested
+	memoryDiff := memoryPctOfCapRequested - avergePctOfCapRequested
+	diff := math.Sqrt(math.Pow(cpuDiff, 2) + math.Pow(memoryDiff, 2))
+	// closer percentage metrics lead to higher score with 10 being the highest
```

Comment:
Hi, maybe I did not make myself very clear...... I googled Manhatten distance and I think what I am suggesting is not.

sqrt(cpu^2 + mem^2) itself cannot reflect how far cpu and mem are from each other, so one would eventually have to use sqrt((cpu-average)^2+(mem-average)^2) to reflect the distance (or without sqrt). Given that average=(cpu + mem)/2, $sqrt((cpu-average)^2+(mem-average)^2)$ and $abs(cpu - mem)$ are in fact both euclidean distance in difference scale, i.e.:

sqrt((cpu-average)^2+(mem-average)^2) = 
sqrt((cpu-(cpu+mem)/2)^2 + (mem-(cpu+mem)/2)^2) = 
sqrt(((cpu-mem)/2)^2 + ((mem-cpu)/2)^2) = 
sqrt((cpu-mem)^2/4) = sqrt((cpu-mem)^2)/2 = 
abs(cpu-mem) / 2

I have submitted a new patch to implement the my ideas (although it needs rebase to pass ci test). By looking at the test cases, I think the function gives reasonble scores for different degree of resource balance.

All in all, I can use sqrt instead of absolute vaule, but as I explained above, they end up the same and what matters only is how to scale them from 0 - 10. Of course, using sqrt may be more intuitive for one to realize that we are considering the distance between resource types.

What do you think?